File: utttcpp/src/nmcts-evaluate.cpp
#include <torch/script.h>
#include <torch/torch.h>

#include <cstdint>
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>

#include "utttcpp/game/constants.hpp"
#include "utttcpp/game/ultimateTicTacToe.hpp"
#include "utttcpp/helpers/prngMersenneTwister.hpp"
#include "utttcpp/selfplay/neuralMonteCarloTreeSearch.hpp"

using utttcpp::X_STATE_VALUE;
using utttcpp::O_STATE_VALUE;
using utttcpp::STATE_SIZE;
using utttcpp::ROW_INDEX;
using utttcpp::COL_INDEX;

using utttcpp::UltimateTicTacToe;

using utttcpp::InitializeMersenneTwisterPRNG;

using utttcpp::NeuralMonteCarloTreeSearch;
using utttcpp::EvaluatedState;
using utttcpp::serializeEvaluatedState;
using utttcpp::EvaluatedAction;
using utttcpp::serializeEvaluatedActions;

int parseInt(char *argv) {
    int number;
    std::stringstream stream(argv);
    stream >> number;
    return number;
}

double parseDouble(char *argv) {
    double number;
    std::stringstream stream(argv);
    stream >> number;
    return number;
}

std::string parseString(char *argv) {
    return std::string(argv);
}

uint32_t parseUInt32(char *argv) {
    uint32_t number;
    std::stringstream stream(argv);
    stream >> number;
    return number;
}

UltimateTicTacToe *parseUttt(char *argv) {
    unsigned char *state = new unsigned char[STATE_SIZE];
    std::string utttState = parseString(argv);
    int stateValue;
    for (int i = 0; i < STATE_SIZE; i++) {
        stateValue = utttState[i] - '0';
        if (stateValue < 0 || stateValue > 9) {
            throw std::runtime_error("invalid state value");
        }
        state[i] = (unsigned char)stateValue;
    }
    UltimateTicTacToe *uttt = new UltimateTicTacToe();
    uttt->setState(state);
    delete[] state;
    return uttt;
}

int main(int argc, char *argv[]) {
    if (argc != 7) {
        throw std::runtime_error(
            "required arguments: "
            "policyValueNetPath, utttState, "
            "numSimulations, explorationStrength, prngSeed, outputPath");
    }

    const std::string policyValueNetPath = parseString(argv[1]);
    std::cout << "policyValueNetPath: " << policyValueNetPath << std::endl;

    std::cout << "utttState: " << argv[2] << std::endl;
    UltimateTicTacToe *uttt = parseUttt(argv[2]);
    std::cout << uttt->toString() << std::endl;

    const int numSimulations = parseInt(argv[3]);
    std::cout << "numSimulations: " << numSimulations << std::endl;

    const double explorationStrength = parseDouble(argv[4]);
    std::cout << "explorationStrength: " << explorationStrength << std::endl;

    const uint32_t prngSeed = parseUInt32(argv[5]);
    std::cout << "prngSeed: " << prngSeed << std::endl;
    InitializeMersenneTwisterPRNG(prngSeed);

    const std::string outputPath = parseString(argv[6]);
    std::cout << "outputPath: " << outputPath << std::endl;

    torch::jit::script::Module policyValueNet;
    policyValueNet = torch::jit::load(policyValueNetPath);
    policyValueNet.to(torch::kCUDA);
    policyValueNet.eval();

    NeuralMonteCarloTreeSearch *nmcts =
        new NeuralMonteCarloTreeSearch(uttt, numSimulations, explorationStrength, policyValueNet);
    nmcts->run();
    std::cout << nmcts->toString() << std::endl;

    EvaluatedState *evaluatedState = nmcts->getEvaluatedState();
    std::string evaluatedStateString = serializeEvaluatedState(evaluatedState);

    EvaluatedAction **evaluatedActions = nmcts->getEvaluatedActions();
    int numEvaluatedActions = nmcts->getNumEvaluatedActions();
    std::string evaluatedActionsString =
        serializeEvaluatedActions(evaluatedActions, numEvaluatedActions);

    std::string evaluationString = evaluatedStateString + " " + evaluatedActionsString;
    std::cout << evaluationString << std::endl;

    std::ofstream ofs(outputPath);
    ofs << evaluationString << std::endl;
    ofs.close();

    std::cout << "evaluation saved to " << outputPath << " successfully!" << std::endl;

    for (int i = 0; i < numEvaluatedActions; i++) {
        delete evaluatedActions[i];
    }
    delete[] evaluatedActions;
    delete[] evaluatedState->state;
    delete evaluatedState;
    delete nmcts;

    return 0;
}

File: scripts/nmcts_evaluate.py
import argparse
import pathlib
import random

import torch
import torch.nn as nn

from utttpy.game.ultimate_tic_tac_toe import UltimateTicTacToe
from utttpy.selfplay.policy_value_network import PolicyValueNetwork
from utttpy.selfplay.neural_monte_carlo_tree_search import (
    NeuralMonteCarloTreeSearch,
    serialize_evaluated_state,
    serialize_evaluated_actions,
)


def run_argparse() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--policy_value_net_path", type=pathlib.Path, required=True)
    parser.add_argument("--uttt_state", type=str, required=True)
    parser.add_argument("--num_simulations", type=int, required=True)
    parser.add_argument("--exploration_strength", type=float, required=True)
    parser.add_argument("--random_seed", type=int, required=True)
    parser.add_argument("--output_path", type=pathlib.Path, required=True)
    parser.add_argument("--device", type=torch.device, default="cuda")
    args = parser.parse_args()
    return args


def load_policy_value_net(state_dict_path: pathlib.Path, device: torch.device) -> nn.Module:
    policy_value_net = PolicyValueNetwork()
    policy_value_net.to(device=device)
    state_dict = torch.load(state_dict_path, map_location=device)
    policy_value_net.load_state_dict(state_dict)
    policy_value_net.eval()
    return policy_value_net


def main() -> None:
    args = run_argparse()
    print(args)

    random.seed(args.random_seed)

    policy_value_net = load_policy_value_net(
        state_dict_path=args.policy_value_net_path,
        device=args.device,
    )

    uttt = UltimateTicTacToe(state=bytearray(map(int, args.uttt_state)))

    nmcts = NeuralMonteCarloTreeSearch(
        uttt=uttt,
        num_simulations=args.num_simulations,
        exploration_strength=args.exploration_strength,
        policy_value_net=policy_value_net,
    )
    nmcts.run(progress_bar=True)
    print(nmcts)

    evaluated_state = nmcts.get_evaluated_state()
    evaluated_actions = nmcts.get_evaluated_actions()

    evaluated_state_str = serialize_evaluated_state(evaluated_state=evaluated_state)
    evaluated_actions_str = serialize_evaluated_actions(evaluated_actions=evaluated_actions)

    evaluation_str = f"{evaluated_state_str} {evaluated_actions_str}"
    print(evaluation_str)

    with open(args.output_path, "w") as f:
        f.write(evaluation_str + "\n")

    print(f"evaluation saved to {args.output_path} successfully!")


if __name__ == "__main__":
    main()
